### PR TITLE
Fix 2D block store offsets

### PR DESCRIPTION
**Pull request by mfrancepillois**
Fix the way the offsets of the `TritonGEN::Matrix2DBlockStoreOp` operation are calculated. 
The 2D block store support now uses a warp stride based method to calculate the replicate's offsets similar to the method used by the 2D load support.

### CODE REVIEW

**Comment by mfrancepillois** [comment]
This bugfix has been tested in the draft PR: https://github.com/intel/intel-xpu-backend-for-triton/pull/1312, which incorporates the full 2D block store support (i.e. the llvm-target branch + https://github.com/intel/intel-xpu-backend-for-triton/pull/1193 + this bugfix).

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp [hunk 1]
```diff
@@ -689,22 +689,31 @@ struct StoreOpConversion
...
     std::array<unsigned, 2> replicaStride = {
-        static_cast<unsigned>(elemsPerInstr[0]),
-        static_cast<unsigned>(elemsPerInstr[1])};
-    std::array<unsigned, 2> warpStride = {
-        static_cast<unsigned>(numReps[0] * elemsPerInstr[0]),
-        static_cast<unsigned>(numReps[1] * elemsPerInstr[1])};
-
-    Value dimWarpId0 = mul(multiDimWarpId[0], i32_val(warpStride[0]));
-    Value dimWarpId1 = mul(multiDimWarpId[1], i32_val(warpStride[1]));
+        (unsigned)(outerDimWarpNum * elemsPerInstr[0]),
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp:703]: use static_cast rather than C style cast

**Review by victor-eds — COMMENTED** [review]
Can we have lit tests for this?

File: third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp (hunk 1 above)
Comment (mfrancepillois) [reply third_party/intel/lib/TritonIntelGPUToLLVM/LoadStoreOpToLLVM.cpp]: The coding style has been improved.

**Comment by mfrancepillois** [comment]
> Can we have lit tests for this?

As most of the updates to the offset calculation are performed in the pass itself and are not in the generated code, I'm not sure what exactly you want to add to the existing test that checks the lowering of `tt.store` to `TritonGEN::Matrix2DBlockStoreOp`
https://github.com/intel/intel-xpu-backend-for-triton/blob/llvm-target/test/TritonIntelGPU/store-to-llvm-2dstore.mlir

**Review by etiotto — APPROVED** [review]

**Review by chengjunlu — APPROVED** [review]

**Review by whitneywhtsang — APPROVED** [review]

**Comment by victor-eds** [comment]
> As most of the updates to the offset calculation are performed in the pass itself and are not in the generated code, I'm not sure what exactly you want to add to the existing test that checks the lowering of `tt.store` to `TritonGEN::Matrix2DBlockStoreOp`

I meant a test case in which this change affects the result. A case which was broken before this change.

**Comment by mfrancepillois** [comment]
> > As most of the updates to the offset calculation are performed in the pass itself and are not in the generated code, I'm not sure what exactly you want to add to the existing test that checks the lowering of `tt.store` to `TritonGEN::Matrix2DBlockStoreOp`
> 
> I meant a test case in which this change affects the result. A case which was broken before this change.

The lit test has been improved to check this change.

**Review by victor-eds — APPROVED** [review]

**Pull request merged**